plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs

```diff
@@ -0,0 +1,251 @@
+  , pluginNotificationHandlers = mconcat
+  [ mkPluginNotificationHandler LSP.STextDocumentDidOpen $
+      \ide vfs _ (DidOpenTextDocumentParams TextDocumentItem{_uri,_version}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocOpened _uri
+        addFileOfInterest ide file Modified{firstOpen=True}
+        restartCabalShakeSession ide vfs file "(opened)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidChange $
+      \ide vfs _ (DidChangeTextDocumentParams VersionedTextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocModified _uri
+        addFileOfInterest ide file Modified{firstOpen=False}
+        restartCabalShakeSession ide vfs file "(changed)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidSave $
+      \ide vfs _ (DidSaveTextDocumentParams TextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocSaved _uri
+        addFileOfInterest ide file OnDisk
+        restartCabalShakeSession ide vfs file "(saved)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidClose $
+      \ide vfs _ (DidCloseTextDocumentParams TextDocumentIdentifier{_uri}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocClosed _uri
+        deleteFileOfInterest ide file
+        restartCabalShakeSession ide vfs file "(closed)"
+  ]
+  }
+  where
+    log' = logWith recorder
+
+    whenUriFile :: Uri -> (NormalizedFilePath -> IO ()) -> IO ()
+    whenUriFile uri act = whenJust (LSP.uriToFilePath uri) $ act . toNormalizedFilePath'
+
+-- | Helper function to restart the shake session, specifically for modifying .cabal files.
+-- No special logic, just group up a bunch of functions you need for the base
+-- Notification Handlers.
+restartCabalShakeSession :: IdeState -> VFS.VFS -> NormalizedFilePath -> String -> IO ()
+restartCabalShakeSession ide vfs file actionMsg = do
+  join $ atomically $ Shake.recordDirtyKeys (shakeExtras ide) GetModificationTime [file]
+  restartShakeSession (shakeExtras ide) (VFSModified vfs) (fromNormalizedFilePath file ++ " " ++ actionMsg) []
+
+-- ----------------------------------------------------------------
+-- Plugin Rules
+-- ----------------------------------------------------------------
+
+data ParseCabal = ParseCabal
+    deriving (Eq, Show, Typeable, Generic)
+instance Hashable ParseCabal
+instance NFData   ParseCabal
+
+type instance RuleResult ParseCabal = ()
+
+cabalRules :: Recorder (WithPriority Log) -> Rules ()
+cabalRules recorder = do
+  -- Make sure we initialise the cabal files-of-interest.
+  ofInterestRules recorder
+  -- Rule to produce diagnostics for cabal files.
+  define (cmapWithPrio LogShake recorder) $ \ParseCabal file -> do
+    t <- use GetModificationTime file
+    log' Debug $ LogModificationTime file t
+    mVirtualFile <- Shake.getVirtualFile file
+    contents <- case mVirtualFile of
```

Comment:
Do we not have a function for this? Seems pretty standard: get the VFS contents or else read from disk.

Reply:
Afaict, we do not, we have 
```haskell
moduleText :: (IsString e, MonadLsp c m) => Uri -> ExceptT e m Text
moduleText uri =
    handleMaybeM "mdlText" $
      (virtualFileText <$>)
          <$> getVirtualFile
              (toNormalizedUri uri)
 ```
 in hls-eval-plugin and 
 ```haskell
     provider :: forall m. FormattingMethod m => SMethod m -> PluginMethodHandler a m
    provider m ide _pid params
      | Just nfp <- uriToNormalizedFilePath $ toNormalizedUri uri = do
        mf <- getVirtualFile $ toNormalizedUri uri
        case mf of
          Just vf -> do
            let typ = case m of
                  STextDocumentFormatting -> FormatText
                  STextDocumentRangeFormatting -> FormatRange (params ^. J.range)
                  _ -> error "mkFormattingHandlers: impossible"
            f ide typ (virtualFileText vf) nfp opts
          Nothing -> pure $ Left $ responseError $ T.pack $ "Formatter plugin: could not get file contents for " ++ show uri
```

in hls-plugin-api. Not in the mood for refactoring that right now.

Reply:
It looks like, most  Core IDE functions don't use the virtual files at all... very confusing...

Reply:
ghcide uses `getFileContentsRule` and `getFileContents` to abstract over `virtualFileText`.

Reply:
It is a bit too complicated to re-use everything, but now it is better.

Reply:
Do you want to quickly note down your findings in an issue or something so we can pick it up later if we want to?

Reply:
In other words, we *do* have a function for it, a rule to be precise. The rule `GetFileContents` abstracts over reading file contents from the virtual file, but we have no rule that alternatively reads the file from disk if it cannot be found on the virtual disk. Not sure that requires an issue.